### PR TITLE
Make `@validate_call` return a function instead of a custom descriptor - fixes binding issue with inheritance and adds `self/cls` argument to validation errors

### PR DESCRIPTION
**Pull request by alexmojaki**
## Change Summary

Testing an alternative to https://github.com/pydantic/pydantic/pull/8258

## Related issue number

Closes https://github.com/pydantic/pydantic/issues/8146

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

File: tests/test_validate_call.py [hunk 1]
```diff
@@ -369,8 +367,8 @@ def foo(self, a: int, b: int):
 
     # insert_assert(exc_info.value.errors(include_url=False))
     assert exc_info.value.errors(include_url=False) == [
-        {'type': 'missing_argument', 'loc': ('a',), 'msg': 'Missing required argument', 'input': ArgsKwargs(())},
-        {'type': 'missing_argument', 'loc': ('b',), 'msg': 'Missing required argument', 'input': ArgsKwargs(())},
+        {'type': 'missing_argument', 'loc': ('a',), 'msg': 'Missing required argument', 'input': ArgsKwargs((x,))},
+        {'type': 'missing_argument', 'loc': ('b',), 'msg': 'Missing required argument', 'input': ArgsKwargs((x,))},
```
Comment (davidhewitt) [inline tests/test_validate_call.py:371]: So the observable change with this approach looks like the error messages will now always show the `self` or `cls` argument. There's debate to be had whether this is a good thing; showing this argument is less implicit but now it doesn't match exactly what the user typed as input arguments, which has potential for confusion.

Comment (alexmojaki) [reply tests/test_validate_call.py:371]: IMO Python already forces users to get used to the idea of the extra self argument with error messages like this:

```python
class C:
  def foo(self, x):
    pass

# TypeError: C.foo() takes 2 positional arguments but 3 were given
C().foo(1, 2)
```

I think there may be times when showing the value of `self` (assuming it has a useful `repr`) might make the error message more useful by providing some extra context about the state of things aside from the actual validation error.

But I'm worried there might be cases where users broadly catch validation errors, serialize them to JSON, and return that in an API. I don't know if that's common but it looks like an intended use case. Then I expect that JSON serialization errors may start appearing, or the validation errors seen by clients (who are far removed from the Python code) will look confusing.

Comment (davidhewitt) [reply tests/test_validate_call.py:371]: Right, good points. I think overall we tend to be ok with error messages changing (we generally justify this by saying that we work to improve error messages). So given that Python itself is a bit mixed here, I'm tempted to argue let's merge this patch as it's a fantastic simplification, and then if we want to work again to improve this later we revisit.

**Comment by sydney-runkle** [comment]
@alexmojaki, ok if I move out of draft so we can merge?

**Review by sydney-runkle — APPROVED** [review]
Nice, great simplification.

**Comment by alexmojaki** [comment]
I think the whole class could be replaced by a function, and all the logic in `_internal/_validate_call.py` could be moved to `validate_call_decorator.py`, but I don't know if it's worth messing with the history.

**Comment by davidhewitt** [comment]
If there's a performance benefit or meaningful simplification to the code, I think that's welcome. Let's merge this one as-is and we can discuss the refactor in a follow-up PR if desired.

**Pull request merged**